x-pack/agent/pkg/agent/transpiler/steps.go

```diff
@@ -0,0 +1,219 @@
+		ks := keys(m)
+		if len(ks) > 1 {
+			return fmt.Errorf("unknown rule identifier, expecting one identifier and received %d", len(ks))
+		}
+
+		name := ks[0]
+		fields := m[name]
+
+		var s Step
+		switch name {
+		case "delete_file":
+			s = &DeleteFileStep{}
+		case "move_file":
+			s = &MoveFileStep{}
+		default:
+			return fmt.Errorf("unknown rule of type %s", name)
+		}
+
+		if err := unpack(fields, s); err != nil {
+			return err
+		}
+
+		steps = append(steps, s)
+	}
+	r.Steps = steps
+	return nil
+}
+
+// DeleteFileStep removes a file from disk.
+type DeleteFileStep struct {
+	Path string
+	// FailOnMissing fails if file is already missing
+	FailOnMissing bool `yaml:"fail_on_missing" config:"fail_on_missing"`
+}
+
+// Execute executes delete file step.
+func (r *DeleteFileStep) Execute(rootDir string) error {
+	path, isSubpath := joinPaths(rootDir, r.Path)
+	if !isSubpath {
+		return fmt.Errorf("invalid path value for operation 'Delete': %s", path)
+	}
+
+	err := os.Remove(path)
+
+	if os.IsNotExist(err) && r.FailOnMissing {
+		// is not found and should be reported
+		return err
+	}
+
+	if err != nil && !os.IsNotExist(err) {
+		// report others
+		return err
+	}
+
+	return nil
+}
+
+// DeleteFile creates a DeleteFileStep
+func DeleteFile(path string, failOnMissing bool) *DeleteFileStep {
+	return &DeleteFileStep{
+		Path:          path,
+		FailOnMissing: failOnMissing,
+	}
+}
+
+// MoveFileStep moves a file to a new location.
+type MoveFileStep struct {
+	Path   string
+	Target string
+	// FailOnMissing fails if file is already missing
+	FailOnMissing bool `yaml:"fail_on_missing" config:"fail_on_missing"`
+}
+
+// Execute executes move file step.
+func (r *MoveFileStep) Execute(rootDir string) error {
+	path, isSubpath := joinPaths(rootDir, r.Path)
+	if !isSubpath {
+		return fmt.Errorf("invalid path value for operation 'Move': %s", path)
+	}
+
+	target, isSubpath := joinPaths(rootDir, r.Target)
+	if !isSubpath {
+		return fmt.Errorf("invalid target value for operation 'Move': %s", target)
+	}
+
+	err := os.Rename(path, target)
+
+	if os.IsNotExist(err) && r.FailOnMissing {
+		// is not found and should be reported
+		return err
+	}
+
+	if err != nil && !os.IsNotExist(err) {
+		// report others
+		return err
+	}
+
+	return nil
+}
+
+// MoveFile creates a MoveFileStep
+func MoveFile(path, target string, failOnMissing bool) *MoveFileStep {
+	return &MoveFileStep{
+		Path:          path,
+		Target:        target,
+		FailOnMissing: failOnMissing,
+	}
+}
+
+// joinPaths joins paths and returns true if path is subpath of rootDir
+func joinPaths(rootDir, path string) (string, bool) {
+	if !filepath.IsAbs(path) {
+		path = filepath.Join(rootDir, path)
+	}
+
+	absRoot := filepath.Clean(filepath.FromSlash(rootDir))
+	absPath := filepath.Clean(filepath.FromSlash(path))
+
+	// path on windows are case insensitive
+	if runtime.GOOS == "windows" {
+		absRoot = strings.ToLower(absRoot)
+		absPath = strings.ToLower(absPath)
+	}
```

Comment:
This is also true on macOS, by default HFS+ is case insensitive. So the following will work.
```
mkdir hello
cd HelLO
```

Reply:
I've been bitten so many times on weird git issues with case insensitive filesystem, the things is if you turn on the case sensitivity on HFS+ it will make many things fails.